src/Data/CAS/RocksDB.hs

```diff
@@ -0,0 +1,410 @@
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- |
+-- Module: Data.CAS.RocksDB
+-- Copyright: Copyright © 2019 Kadena LLC.
+-- License: MIT
+-- Maintainer: Lars Kuhtz <lars@kadena.io>
+-- Stability: experimental
+--
+-- TODO
+--
+module Data.CAS.RocksDB
+( RocksDb(..)
+, rocksDbHandle
+, rocksDbNamespace
+, openRocksDb
+, closeRocksDb
+, withRocksDb
+, withTempRocksDb
+
+-- * Rocks DB Table
+, Codec(..)
+, RocksDbTable
+, newTable
+, tableLookup
+, tableInsert
+, tableDelete
+
+-- * Rocks DB Table Iterator
+, RocksDbTableIter
+, createTableIter
+, releaseTableIter
+, withTableIter
+, tableIterValid
+
+-- ** Seeking
+, tableIterSeek
+, tableIterFirst
+, tableIterLast
+
+-- ** Advance Iterator
+, tableIterNext
+, tableIterPrev
+
+-- ** Query Iterator
+, tableIterEntry
+, tableIterValue
+, tableIterKey
+
+-- ** Streams
+, iterToEntryStream
+, iterToValueStream
+, iterToKeyStream
+
+-- * RocksDbCas
+, RocksDbCas(..)
+, newCas
+) where
+
+import Control.Lens
+import Control.Monad.Catch
+import Control.Monad.IO.Class
+
+import qualified Data.ByteString as B
+import Data.CAS
+import qualified Data.Text as T
+
+import qualified Database.RocksDB.Base as R
+import qualified Database.RocksDB.Iterator as I
+
+import GHC.Generics (Generic)
+
+import qualified Streaming.Prelude as S
+
+import System.IO.Temp
+
+-- internal modules
+import Chainweb.Utils hiding (Codec)
+
+-- -------------------------------------------------------------------------- --
+-- RocksDb
+
+-- | This wrapper allows to create key namespaces. This is, for instance, useful
+-- for testing when running several concurrent tests on the same rocks db
+-- instances.
+--
+data RocksDb = RocksDb
+    { _rocksDbHandle :: !R.DB
+    , _rocksDbNamespace :: !B.ByteString
+    }
+
+makeLenses ''RocksDb
+
+openRocksDb :: FilePath -> IO RocksDb
+openRocksDb path = RocksDb <$> R.open path opts <*> mempty
+  where
+    opts = R.defaultOptions { R.createIfMissing = True }
+
+closeRocksDb :: RocksDb -> IO ()
+closeRocksDb = R.close . _rocksDbHandle
+
+withRocksDb :: FilePath -> (RocksDb -> IO a) -> IO a
+withRocksDb path = bracket (openRocksDb path) closeRocksDb
+
+withTempRocksDb :: String -> (RocksDb -> IO a) -> IO a
+withTempRocksDb template f = withSystemTempDirectory template $ \dir ->
+    withRocksDb dir f
+
+-- -------------------------------------------------------------------------- --
+-- RocksDb Table
+
+-- NOTE: the implementation currently doesnt' support iteration over nested
+-- namespaces. This could easily be added via an iterator that iterates over
+-- all respective prefixes ending with '/' or '$' and using a codec that
+-- could handle the types in all sub-namespaces. This could for instance
+-- be useful for iterating over the complete database.
+
+-- | Encodings of keys must NOT contain any of '$', '%', '/', and '0'. This is
```

Comment:
I tried to keep it simple for now. But, yes, there is no reason that the code doesn't check it, except for me being lazy and trying to get something working quickly.

Generally, I think, this is a somewhat standard technique to encode namespaces in key value stores when the suffix of the keys can be arbitrary bytes. At least I've seen similar encodings being used elsewhere.

Modern versions of RocksDB support a more direct encoding of key namespaces. I think "Column Families" would be useful here, but that's not supported by the Haskell API.

Also, it could be beneficial to have an implementation that only uses more standard low-level features of RocksDB, because it would limit our the dependency on it. It would make it easier to move to another backend (say, level db) in case RocksDB doesn't meet our expectations. It also makes it easy to implement an in memory backend for testing.

Reply:
I'll add a commit that add the checks.

Reply:
Oops, I just noticed that this comment is miss-placed. The restriction applies only for table names. Actually, the reason for these restriction on table names is so that keys don't require any restriction.

Sorry, for the confusion. I'll fix it.